src/ray/gcs/redis_context-inl.h

```diff
@@ -0,0 +1,159 @@
+    } else if (newContext->err) {
+      oss << "Could not establish connection to Redis " << address << ":" << port
+          << " (context.err = " << newContext->err << ").";
+    }
+    return std::make_pair(Status::RedisError(oss.str()), nullptr);
+  }
+  return std::make_pair(Status::OK(),
+                        std::unique_ptr<RedisContextType, RedisContextDeleter>(
+                            newContext, RedisContextDeleter()));
+}
+
+template <typename RedisContextType, typename RedisConnectFunctionType>
+std::pair<Status, std::unique_ptr<RedisContextType, RedisContextDeleter>>
+ConnectWithRetries(const std::string &address,
+                   int port,
+                   const RedisConnectFunctionType &connect_function) {
+  RAY_LOG(INFO) << "Attempting to connect to address " << address << ":" << port << ".";
+  int connection_attempts = 0;
+  auto resp = ConnectWithoutRetries<RedisContextType>(address, port, connect_function);
+  auto status = resp.first;
+  while (!status.ok()) {
+    if (connection_attempts >= RayConfig::instance().redis_db_connect_retries()) {
+      RAY_LOG(FATAL) << RayConfig::instance().redis_db_connect_retries() << " attempts "
+                     << "to connect have all failed. Please check whether the"
+                     << " redis storage is alive or not. The last error message was: "
+                     << status.ToString();
+      break;
+    }
+    RAY_LOG_EVERY_MS(ERROR, 1000)
+        << "Failed to connect to Redis due to: " << status.ToString()
+        << ". Will retry in "
+        << RayConfig::instance().redis_db_connect_wait_milliseconds() << "ms.";
+
+    // Sleep for a little.
+    std::this_thread::sleep_for(std::chrono::milliseconds(
+        RayConfig::instance().redis_db_connect_wait_milliseconds()));
+    resp = ConnectWithoutRetries<RedisContextType>(address, port, connect_function);
+    status = resp.first;
+    connection_attempts += 1;
+  }
+  return resp;
+}
+
+RedisAsyncContext *CreateAsyncContext(
+    std::unique_ptr<redisAsyncContext, RedisContextDeleter> context) {
+  return new RedisAsyncContext(std::move(context));
+}
+
+template <typename RedisContextType>
+void RedisResponseFn(struct redisAsyncContext *async_context,
+                     void *raw_reply,
+                     void *privdata) {
+  auto *request_cxt = static_cast<RedisRequestContext *>(privdata);
+  auto redis_reply = reinterpret_cast<redisReply *>(raw_reply);
+  // Error happened.
+  if (redis_reply == nullptr || redis_reply->type == REDIS_REPLY_ERROR) {
+    auto error_msg = redis_reply ? redis_reply->str : async_context->errstr;
+    RAY_LOG(ERROR) << "Redis request [" << absl::StrJoin(request_cxt->redis_cmds_, " ")
+                   << "]"
+                   << " failed due to error " << error_msg << ". "
+                   << request_cxt->pending_retries_ << " retries left.";
+
+    // Retry the request after a while.
+    auto delay = request_cxt->exp_back_off_.Current();
```

Comment:
Seems we should always call `request_cxt->exp_back_off_.Next();`, I think we should just call it here unconditionally.